jax/lib/xla_bridge.py

```diff
@@ -110,61 +113,80 @@ def get_compile_options(num_replicas, num_partitions, device_assignment=None):
-  if not platform:
-    platform = FLAGS.jax_platform_name or None
 
+# TODO(skye): remove this function once xla_client.get_local_backend() never
+# returns None.
+def _get_local_backend(platform):
   backend = xla_client.get_local_backend(platform)
   if backend is None:
     raise RuntimeError("No local XLA backends found.")
-
-  if backend.platform == 'cpu' and platform != 'cpu':
-    warnings.warn('No GPU/TPU found, falling back to CPU.')
-
   return backend
 
+register_backend('cpu', lambda unused_platform=None: _get_local_backend("cpu")) # type: ignore
+register_backend('gpu', lambda unused_platform=None: _get_local_backend("gpu")) # type: ignore
 
-register_backend('xla', _get_local_backend)
 
-# memoize the TPU driver to be consistent with xla_client behavior
-_tpu_backend = None
+def _get_tpu_driver_backend(unused_platform=None):
+  backend_target = FLAGS.jax_backend_target
+  if backend_target is None:
+    raise ValueError('When using TPU Driver as the backend, you must specify '
+                     '--jax_backend_target=<hostname>:8470.')
+  return tpu_client.TpuBackend.create(worker=backend_target)
 
-def _get_tpu_driver_backend(platform):
-  del platform
-  global _tpu_backend
-  if _tpu_backend is None:
-    backend_target = FLAGS.jax_backend_target
-    if backend_target is None:
-      raise ValueError('When using TPU Driver as the backend, you must specify '
-                       '--jax_backend_target=<hostname>:8470.')
-    _tpu_backend = tpu_client.TpuBackend.create(worker=backend_target)
-  return _tpu_backend
-
-
-if tpu_client:
-  register_backend('tpu_driver', _get_tpu_driver_backend)
+if tpu_client and FLAGS.jax_xla_backend == 'tpu_driver':
+  register_backend('tpu', _get_tpu_driver_backend)
 
 
+_backends = None
 _backend_lock = threading.Lock()
 
 @util.memoize
-def get_backend(platform=None):
+def get_backend(platform: Optional[str] = None) -> LocalClient:
   # TODO(mattjj,skyewm): remove this input polymorphism after we clean up how
   # 'backend' values are handled
   if not isinstance(platform, (type(None), str)):
     return platform
 
   with _backend_lock:
-    backend = _backends.get(FLAGS.jax_xla_backend)
-    if backend is None:
-      msg = 'Unknown jax_xla_backend value "{}".'
-      raise ValueError(msg.format(FLAGS.jax_xla_backend))
-    return backend(platform)
+    if _backends is None:
+      _initialize_backends()
+  assert _backends is not None
+
+  if platform is None:
+    return list(_backends.values())[-1]
+  if platform not in _backends:
+    raise ValueError(f"Unknown backend '{platform}'")
+  return _backends[platform]
+
+
+def _initialize_backends():
+  global _backends
+  assert _backends is None, "_initialize_backends() should only be called once!"
+  _backends = OrderedDict()
+  for name, factory in _backend_factories.items():
+    logging.vlog(1, f"Initializing backend '{name}'")
+    try:
+      # TODO(skye): remove name parameter once all backend factories create a
```

Comment:
Should this just be:
```
# TODO(skye): remove the name parameter
```
because there's no condition on it, it just needs doing now.